Make getKeyLength return an optional

nmahendru opened the pull request:
#### c29b8687ec88543d82a462ace9b3eef38d71024f
<pre>
Make getKeyLength return an optional
<a href="https://bugs.webkit.org/show_bug.cgi?id=282096">https://bugs.webkit.org/show_bug.cgi?id=282096</a>
<a href="https://rdar.apple.com/138656378">rdar://138656378</a>

Reviewed by Matthew Finkel and David Kilzer.

WebKit&apos;s current implementation has no way to return `null` for getKeyLength.
It always returns a size_t which will always have a value.

But the ability to return an optional(no value meaning null) is needed for adherence to the spec for the below algorithms.
More details:
[1] <a href="https://w3c.github.io/webcrypto/#pbkdf2-operations">https://w3c.github.io/webcrypto/#pbkdf2-operations</a>
[2] <a href="https://w3c.github.io/webcrypto/#hkdf-operations">https://w3c.github.io/webcrypto/#hkdf-operations</a>

Also this is needed as a first step for re-landing <a href="https://commits.webkit.org/285383@main.">https://commits.webkit.org/285383@main.</a>
285383@main exposed a problem wherein the deriveKey output changed without any change on client side code.
deriveKey calls getKeyLength internally and that is hardcoded to zero(seemingly for representing null) for HKDF/PBKDF2.
But that is fundamentally wrong.
With 285383@main we started treating zero length as zero length for deriveBits.
deriveKey basically takes the deriveKeyAlgorithm and calls getKeylength on that.
Then it calls deriveBits using that length. Before 28583@main our ECDH implementation
was returning the whole key on zero length for deriveBits, so deriveKey just worked fine.
But with 285383@main, that length was &quot;correctly&quot; used by the internal deriveBits operation to truncate to zero
and the deriveKey output thus changed completely.
So any use of the derivedKey will also produce a &quot;different&quot; output.

The test added here expands the scope of deriveKey tests that we have so that the derivedKey is actually exercised
and a scenario where two parties are trying to derive the same key actually works.

* LayoutTests/crypto/subtle/derive-key-for-all-supported-algorithms-expected.txt: Added.
* LayoutTests/crypto/subtle/derive-key-for-all-supported-algorithms.html: Added.
* Source/WebCore/crypto/CryptoAlgorithm.cpp:
(WebCore::CryptoAlgorithm::getKeyLength):
* Source/WebCore/crypto/CryptoAlgorithm.h:
* Source/WebCore/crypto/algorithms/CryptoAlgorithmAESCBC.cpp:
(WebCore::CryptoAlgorithmAESCBC::getKeyLength):
* Source/WebCore/crypto/algorithms/CryptoAlgorithmAESCBC.h:
* Source/WebCore/crypto/algorithms/CryptoAlgorithmAESCFB.cpp:
(WebCore::CryptoAlgorithmAESCFB::getKeyLength):
* Source/WebCore/crypto/algorithms/CryptoAlgorithmAESCFB.h:
* Source/WebCore/crypto/algorithms/CryptoAlgorithmAESCTR.cpp:
(WebCore::CryptoAlgorithmAESCTR::getKeyLength):
* Source/WebCore/crypto/algorithms/CryptoAlgorithmAESCTR.h:
* Source/WebCore/crypto/algorithms/CryptoAlgorithmAESGCM.cpp:
(WebCore::CryptoAlgorithmAESGCM::getKeyLength):
* Source/WebCore/crypto/algorithms/CryptoAlgorithmAESGCM.h:
* Source/WebCore/crypto/algorithms/CryptoAlgorithmAESKW.cpp:
(WebCore::CryptoAlgorithmAESKW::getKeyLength):
* Source/WebCore/crypto/algorithms/CryptoAlgorithmAESKW.h:
* Source/WebCore/crypto/algorithms/CryptoAlgorithmHKDF.cpp:
(WebCore::CryptoAlgorithmHKDF::getKeyLength):
* Source/WebCore/crypto/algorithms/CryptoAlgorithmHKDF.h:
* Source/WebCore/crypto/algorithms/CryptoAlgorithmHMAC.cpp:
(WebCore::CryptoAlgorithmHMAC::getKeyLength):
* Source/WebCore/crypto/algorithms/CryptoAlgorithmHMAC.h:
* Source/WebCore/crypto/algorithms/CryptoAlgorithmPBKDF2.cpp:
(WebCore::CryptoAlgorithmPBKDF2::getKeyLength):
* Source/WebCore/crypto/algorithms/CryptoAlgorithmPBKDF2.h:
* Source/WebCore/crypto/keys/CryptoKeyAES.cpp:
(WebCore::CryptoKeyAES::getKeyLength):
* Source/WebCore/crypto/keys/CryptoKeyAES.h:
* Source/WebCore/crypto/keys/CryptoKeyHMAC.cpp:
(WebCore::getKeyLengthFromHash):
(WebCore::CryptoKeyHMAC::generate):
(WebCore::CryptoKeyHMAC::getKeyLength):
* Source/WebCore/crypto/keys/CryptoKeyHMAC.h:

Canonical link: <a href="https://commits.webkit.org/285769@main">https://commits.webkit.org/285769@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3aef888a3ff465a52a7c7f611914449d4702c434

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/73769 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/53198 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/26580 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/78091 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/25007 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/75886 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/62331 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/983 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/58030 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/16397 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/76836 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/48169 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/63472 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/38429 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/44958 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/20960 "Passed tests") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/23340 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/66529 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/21307 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/79658 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/1086 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/541 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/66365 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/1229 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/63482 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/65644 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/9511 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/7695 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11374 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/1050 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/3800 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/1079 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/1066 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/1085 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

ddkilzer commented on Source/WebCore/crypto/algorithms/CryptoAlgorithmHKDF.cpp:
```diff
@@ -75,9 +75,9 @@ void CryptoAlgorithmHKDF::importKey(CryptoKeyFormat format, KeyData&& data, cons
+ExceptionOr<std::optional<size_t>> CryptoAlgorithmHKDF::getKeyLength(const CryptoAlgorithmParameters&)
 {
-    return 0;
+    return std::optional<size_t>();
```
Nit:  More idiomatic in WebKit to do this:

```suggestion
    return { };
```

nmahendru replied:
I tried that first. It does not compile. Exception or only has rvalue constructors.

ddkilzer commented on Source/WebCore/crypto/algorithms/CryptoAlgorithmPBKDF2.cpp:
```diff
@@ -74,9 +74,9 @@ void CryptoAlgorithmPBKDF2::importKey(CryptoKeyFormat format, KeyData&& data, co
+ExceptionOr<std::optional<size_t>> CryptoAlgorithmPBKDF2::getKeyLength(const CryptoAlgorithmParameters&)
 {
-    return 0;
+    return std::optional<size_t>();
```
```suggestion
    return { };
```

sysrqb commented on LayoutTests/crypto/subtle/derive-key-for-all-supported-algorithms.html:
```diff
@@ -0,0 +1,97 @@
+<script type="text/javascript">
+description("We test deriveKey and then make sure that ecdh produces the same shared secret and the same encryption results using a key derived from that secret.");
+
+/* generating data
```
Not needed anymore or is this for documentation purposes? If it's for documenting how you got the values, then I'd suggest adding a few more words.

nmahendru replied:
added

ddkilzer commented on Source/WebCore/crypto/keys/CryptoKeyHMAC.cpp:
```diff
@@ -46,8 +46,7 @@ static size_t getKeyLengthFromHash(CryptoAlgorithmIdentifier hash)
     default:
-        ASSERT_NOT_REACHED();
-        return 0;
+        return std::nullopt;
```
I think this is equivalent—please change all of these to be the same (either `{ }` or `std::nullopt`) across all the files:

```suggestion
        return { };
```

ddkilzer commented on Source/WebCore/crypto/keys/CryptoKeyHMAC.cpp:
```diff
@@ -126,15 +126,14 @@ JsonWebKey CryptoKeyHMAC::exportJwk() const
-        return result;
-
-    return Exception { ExceptionCode::TypeError };
+    std::optional<size_t> result = aesParameters.length ? *(aesParameters.length) : getKeyLengthFromHash(aesParameters.hashIdentifier);
```
Would be nice to rename `result` to `length` to match the code above since we're making a change here.

sysrqb replied:
Also could use `auto` instead of `std::optional<size_t>`.

ddkilzer reviewed (approved):
r=me, but please consider making the requested changes.

sysrqb commented on Source/WebCore/crypto/keys/CryptoKeyHMAC.cpp:
```diff
@@ -46,8 +46,7 @@ static size_t getKeyLengthFromHash(CryptoAlgorithmIdentifier hash)
     case CryptoAlgorithmIdentifier::SHA_512:
         return 1024;
     default:
-        ASSERT_NOT_REACHED();
```
Should this be reachable?

nmahendru replied:
It should not. I have added a RELEASE_ASSERT now.

nmahendru committed:
Make getKeyLength return an optional
https://bugs.webkit.org/show_bug.cgi?id=282096
rdar://138656378

Reviewed by Matthew Finkel and David Kilzer.

WebKit's current implementation has no way to return `null` for getKeyLength.
It always returns a size_t which will always have a value.

But the ability to return an optional(no value meaning null) is needed for adherence to the spec for the below algorithms.
More details:
[1] https://w3c.github.io/webcrypto/#pbkdf2-operations
[2] https://w3c.github.io/webcrypto/#hkdf-operations

Also this is needed as a first step for re-landing https://commits.webkit.org/285383@main.
285383@main exposed a problem wherein the deriveKey output changed without any change on client side code.
deriveKey calls getKeyLength internally and that is hardcoded to zero(seemingly for representing null) for HKDF/PBKDF2.
But that is fundamentally wrong.
With 285383@main we started treating zero length as zero length for deriveBits.
deriveKey basically takes the deriveKeyAlgorithm and calls getKeylength on that.
Then it calls deriveBits using that length. Before 28583@main our ECDH implementation
was returning the whole key on zero length for deriveBits, so deriveKey just worked fine.
But with 285383@main, that length was "correctly" used by the internal deriveBits operation to truncate to zero
and the deriveKey output thus changed completely.
So any use of the derivedKey will also produce a "different" output.

The test added here expands the scope of deriveKey tests that we have so that the derivedKey is actually exercised
and a scenario where two parties are trying to derive the same key actually works.

* LayoutTests/crypto/subtle/derive-key-for-all-supported-algorithms-expected.txt: Added.
* LayoutTests/crypto/subtle/derive-key-for-all-supported-algorithms.html: Added.
* Source/WebCore/crypto/CryptoAlgorithm.cpp:
(WebCore::CryptoAlgorithm::getKeyLength):
* Source/WebCore/crypto/CryptoAlgorithm.h:
* Source/WebCore/crypto/algorithms/CryptoAlgorithmAESCBC.cpp:
(WebCore::CryptoAlgorithmAESCBC::getKeyLength):
* Source/WebCore/crypto/algorithms/CryptoAlgorithmAESCBC.h:
* Source/WebCore/crypto/algorithms/CryptoAlgorithmAESCFB.cpp:
(WebCore::CryptoAlgorithmAESCFB::getKeyLength):
* Source/WebCore/crypto/algorithms/CryptoAlgorithmAESCFB.h:
* Source/WebCore/crypto/algorithms/CryptoAlgorithmAESCTR.cpp:
(WebCore::CryptoAlgorithmAESCTR::getKeyLength):
* Source/WebCore/crypto/algorithms/CryptoAlgorithmAESCTR.h:
* Source/WebCore/crypto/algorithms/CryptoAlgorithmAESGCM.cpp:
(WebCore::CryptoAlgorithmAESGCM::getKeyLength):
* Source/WebCore/crypto/algorithms/CryptoAlgorithmAESGCM.h:
* Source/WebCore/crypto/algorithms/CryptoAlgorithmAESKW.cpp:
(WebCore::CryptoAlgorithmAESKW::getKeyLength):
* Source/WebCore/crypto/algorithms/CryptoAlgorithmAESKW.h:
* Source/WebCore/crypto/algorithms/CryptoAlgorithmHKDF.cpp:
(WebCore::CryptoAlgorithmHKDF::getKeyLength):
* Source/WebCore/crypto/algorithms/CryptoAlgorithmHKDF.h:
* Source/WebCore/crypto/algorithms/CryptoAlgorithmHMAC.cpp:
(WebCore::CryptoAlgorithmHMAC::getKeyLength):
* Source/WebCore/crypto/algorithms/CryptoAlgorithmHMAC.h:
* Source/WebCore/crypto/algorithms/CryptoAlgorithmPBKDF2.cpp:
(WebCore::CryptoAlgorithmPBKDF2::getKeyLength):
* Source/WebCore/crypto/algorithms/CryptoAlgorithmPBKDF2.h:
* Source/WebCore/crypto/keys/CryptoKeyAES.cpp:
(WebCore::CryptoKeyAES::getKeyLength):
* Source/WebCore/crypto/keys/CryptoKeyAES.h:
* Source/WebCore/crypto/keys/CryptoKeyHMAC.cpp:
(WebCore::getKeyLengthFromHash):
(WebCore::CryptoKeyHMAC::generate):
(WebCore::CryptoKeyHMAC::getKeyLength):
* Source/WebCore/crypto/keys/CryptoKeyHMAC.h:

Canonical link: https://commits.webkit.org/285769@main
